File: crates/dewasm-backend-go/units/wasi/fd_allocate.go
// fd_allocate grows the file so at least offset+len bytes are backed; it never shrinks (a request within the current size is a no-op).
// Emulated with
// Truncate, which is enough for the guests we target.
func (w *WASI) wasi_fd_allocate(fd uint32, offset, length uint64) uint32 {
    f, ok := w.fds[fd].(*os.File)
    if !ok {
        return wasiBadf
    }
    fi, err := f.Stat()
    if err != nil {
        return wasiIo
    }
    if need := int64(offset + length); need > fi.Size() {
        if err := f.Truncate(need); err != nil {
            return wasiIo
        }
    }
    return wasiOk
}

File: crates/dewasm-backend-ruby/units/wasi/fd_allocate.rb
def wasi_fd_allocate(fd, offset, len)
  io = @fds[fd]
  return ERRNO_BADF if io.nil? || io.is_a?(WasiDir)
  # fd_allocate only ever grows the file to at least offset+len; a request that ends within the current size is a no-op.
  needed = offset + len
  io.truncate(needed) if needed > io.size
  ERRNO_SUCCESS
rescue SystemCallError
  ERRNO_IO
end
